static/js/widgetize.js

```diff
@@ -315,6 +315,153 @@ widgets.poll = (function () {
+
+    self.trivia_question = {};
+    var correct_answer;
+
+    self.set_trivia_ques = function (result) {
+        var question = result.results[0].question;
+
+        correct_answer = result.results[0].correct_answer;
+
+        var options = result.results[0].incorrect_answers;
+        options.push(correct_answer);
+        self.trivia_question  =  {
+            question : question,
+            options : _.shuffle(options),
+            correct_answer :  correct_answer,
+        };
+    };
+
+    self.promise = $.ajax({
+            url: "https://opentdb.com/api.php?amount=1&type=multiple",
+        });
+
+    self.handle = {
+        option_click: {
+            outbound: function (option) {
+                var event = {
+                    type: 'option_click',
+                    option: option,
+                    trivia_question : self.trivia_question,
+                    correct_answer : correct_answer,
+                };
+                return event;
+            },
+
+            inbound: function (sender_id, data) {
+                self.trivia_question = data.trivia_question;
+                correct_answer = $("<div/>").html(data.correct_answer).text();
+                if (data.option===correct_answer && !_.has(self.has_answered,sender_id)) {
+                    self.correctly_answered = true;
+                    self.msg = "Correct Answer: " + people.get_full_name(sender_id);
+                } else {
+                    self.correctly_answered = false;
+                    self.msg = "Wrong Answer: " + people.get_full_name(sender_id);
+                }
+                self.has_answered[sender_id] = true;
+            },
+        },
+        set_question: {
+            outbound: function () {
+                var event = {
+                    type: 'set_question',
+                    trivia_question: self.trivia_question,
+                    correct_answer: correct_answer,
+                };
+                return event;
+            },
+            inbound: function (sender_id, data) {
+                self.trivia_question = data.trivia_question;
+                correct_answer = data.correct_answer;
+            },
+        },
+    };
+
+    self.handle_event = function (sender_id, data) {
+        var type = data.type;
+        if (self.handle[type]) {
+            self.handle[type].inbound(sender_id, data);
+        }
+    };
+
+    return self;
+};
+
+widgets.trivia = (function () {
+    var cls = {};
+
+    cls.activate = function (opts) {
+        var self = {};
+
+        var elem = opts.elem;
+        var callback = opts.callback;
+
+        var trivia_data = trivia_data_holder();
+
+        function render() {
+            if (trivia_data.correctly_answered) {
+                elem.html(trivia_data.msg);
+            } else {
+                if (_.isEmpty(trivia_data.trivia_question)) {
+                    elem.html("Loading...");
+                    trivia_data.promise.done(function (result) {
```

Comment:
I think we can avoid the need for the `promise` if we just take the approach that we use all the data that we've been given so far to determine what we want to render, and we only render when we've processed all the necessary data.  You can see this is how the poll widget works.

Reply:
I cannot figure out a way of doing so, in poll we instantly have the data when user submits his response, that difference is causing all the problems. Can you elaborate?